compiler/ast/dag/keyfilter.go

```diff
@@ -0,0 +1,108 @@
+package dag
+
+import (
+	"github.com/brimdata/zed/order"
+	"github.com/brimdata/zed/pkg/field"
+)
+
+type KeyFilter struct {
+	Expr Expr
+}
+
+// NewKeyFilter creates a KeyFilter that contains a modified form of node where
+// only predicates operating on key are kept. The underlying expression is
+// rewritten in a manner so that results produced by the filter will always be
+// a superset of the results produced by the parent filter (i.e., it will not
+// filter values that would not be not also filtered by the original filter).
+// Currently KeyFilter only recognizes simple key predicates against a literal
+// value and using the comparators ==, >=, >, <, and <=; otherwise the predicate is
+// ignored.
+func NewKeyFilter(key field.Path, node Expr) *KeyFilter {
+	e := visitLeaves(node, func(cmp string, lhs *This, rhs *Literal) Expr {
+		if !key.Equal(lhs.Path) {
+			return nil
+		}
+		return NewBinaryExpr(cmp, lhs, rhs)
```

Comment:
Ugh there's an issue here on a filter: `poolkey == "foo" or notpoolkey == "bar"`. This will get optimized but should not as it could result in false negatives (unless `notpoolkey` is indexed but that's a situation we do not handle).